pkg/commands/drift/drift.go

```diff
@@ -79,21 +93,51 @@ func Process(ctx context.Context, organizationID, bucketQuery, driftignoreFile s
 		return nil, fmt.Errorf("failed to expand graph for ignored assets: %w", err)
 	}
 
-	gcpIAM, err := actualGCPIAM(ctx, organizationID, foldersByID, projectsByID)
+	g, gCtx := errgroup.WithContext(ctx)
+
+	logger.Debugw("fetching iam for org, folders and projects",
+		"number_of_folders", len(folders),
+		"number_of_projects", len(projects))
+	gcpIAMCount := 1 + len(foldersByID) + len(projectsByID)
+	gcpIAMChan := make(chan []*iam.AssetIAM, gcpIAMCount)
+	err = actualGCPIAM(gCtx, g, organizationID, foldersByID, projectsByID, gcpIAMChan)
```

Comment:
I'm curious why this function uses a different concurrency style than the previous (this one uses channels, the other one assigns to a local variable from a goroutine). The other one seemed cleaner to me, since it avoids all the complexity of channel buffering and such, but that's just my opinion. Could `gcpIAM` and `tfIAM` just be return values rather than being read from a channel? You always have the option of creating new/separate nested errgroups inside terraformStateIAM and actualGCPIAM.

Likely I'm just missing something, but it's worth discussing.

Reply:
There were a few things driving my design here:

1) I wanted to fully parallelize the GCP IAM and Terraform IAM operations so they happened concurrently. This meant that I couldn't fully encapsulate the waitGroup/errorGroup in their methods. (I wish there was a `Promise.All()` equivalent for combining waits from multiple goroutines/waitGroups).
2) Use of channels here is required because we have concurrent writes to the same collection (map). I could instead use a mutex and lock before mutating the collection. IMO this might be a bit cleaner since it avoids the translation back from channels to the data format we want.
3) We didn't have to use a mutex on the first 3 operations since they aren't mutating the same collections. Its 1 operation per expected output.

I could take the hit so that we don't parallelize the TF IAM and GCP IAM ops - I do think this might be a little cleaner.

Reply:
Those reasons sound good to me. I'm fine with leaving it as is, or rewriting per your idea in your last paragraph.